### PR TITLE
_WD_FrameList - added Frame visibility info

### PR DESCRIPTION
**Pull request by mlipok**
### Proposed changes

_WD_FrameList should return information about visibility of the frame

for example:
https://stackoverflow.com/questions/19669786/check-if-element-is-visible-in-dom

image

### Checklist

- [x] I have read and noticed the [CODE OF CONDUCT](https://github.com/Danp2/au3WebDriver/blob/master/docs/CODE_OF_CONDUCT.md) document
- [x] I have read and noticed the [CONTRIBUTING](https://github.com/Danp2/au3WebDriver/blob/master/docs/CONTRIBUTING.md) document
- [ ] I have added necessary documentation or screenshots (if appropriate)

### Types of changes

Please check `x` the type of change your PR introduces:

- [ ] Bugfix (change which fixes an issue)
- [x] Feature (change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (functional, structural)
- [ ] Documentation content changes
- [ ] Other (please describe)

### What is the current behavior?
No information about FrameVisibility

### What is the new behavior?

added `_WD_FRAMELIST_FrameVisibility` and new column `IsHidden` to the resulted array 

image

### Additional context

was discussed here: https://github.com/Danp2/au3WebDriver/issues/388

### System under test

not relatead

### CODE REVIEW

File: wd_helper.au3
Comment (Danp2) [inline wd_helper.au3:868]: Have you tested this? AFAICS, this won't be accurate when a parent container is hidden and not the actual frame.

Comment (mlipok) [reply wd_helper.au3:868]: If parent container is hidden, do you expect that actual fram will be visible?

Comment (mlipok) [reply wd_helper.au3:868]: Take a look on 
https://user-images.githubusercontent.com/11089482/186519722-2d7d707a-e1d4-49a4-af2b-92a05e6c4595.png

Is this the exact case you are refering to ?

Comment (Danp2) [reply wd_helper.au3:868]: Yes, that is one instance. But you can't just check the parent element if that's what you are thinking.

Comment (mlipok) [reply wd_helper.au3:868]: > Take a look on https://user-images.githubusercontent.com/11089482/186519722-2d7d707a-e1d4-49a4-af2b-92a05e6c4595.png
> 
> Is this the exact case you are refering to ?

IMHO this case should show 'IsHidden' status as true.

Comment (mlipok) [reply wd_helper.au3:868]: > But you can't just check the parent element if that's what you are thinking.

No this is not my idea.

Comment (Danp2) [reply wd_helper.au3:868]: > IMHO this case should show 'IsHidden' status as true.

Correct... I didn't mean to imply otherwise

> No this is not my idea.

ok

Comment (mlipok) [reply wd_helper.au3:868]: So I do not understand.

Comment (mlipok) [reply wd_helper.au3:868]: > this won't be accurate when a parent container is hidden and not the actual frame.

What you are refering here ?

Comment (Danp2) [reply wd_helper.au3:868]: >>   this won't be accurate when a parent container is hidden and not the actual frame.

> What you are refering here ?

In this case, `a parent container` refers to the parent element *and* any of its ancestral parent elements. This is why I said "But you can't just check the parent element if that's what you are thinking" because simply checking the parentElement's visibility isn't sufficient to ensure that the frame is visible.

Comment (Danp2) [reply wd_helper.au3:868]: You may want to test the `Displayed` functionality of _WD_ElementAction to see if it works correctly for this.

Comment (mlipok) [reply wd_helper.au3:868]: Do you mean that we should check follow this: https://www.w3.org/TR/webdriver/#element-displayedness
`The visibility of an [element](https://www.w3.org/TR/webdriver/#dfn-elements) is guided by what is perceptually visible to the human eye. In this context, an [element](https://www.w3.org/TR/webdriver/#dfn-elements)’s displayedness does not relate to the [visibility](https://www.w3.org/TR/webdriver/#dfn-visibility) or [display](https://www.w3.org/TR/webdriver/#dfn-display) style properties.`

and use `_WD_ElementAction` with `DISPLAYED` instead checking it via `HTML DOM` ?

Comment (mlipok) [reply wd_helper.au3:868]: Does this mean that element that is visible in `HTML DOM` but is not visible to the human eyes because it is on the 3rd page and to make it visible current view should be scroled down.... is treated differently?

differently... when I use 

- `HTML DOM` which should say is not hidden
- WEBDRIVER which should say is not visible to human eye, as this is far, far away on the bottom of the page.

?

**Comment by mlipok** [comment]
Working on `_WD_ElementAction` with `DISPLAYED`

As for now I have:
image

**Comment by mlipok** [comment]
I get into a issue with getting value:
```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/422c9a63-392f-4651-b0ac-ec2069ab9a45/execute/sync; Data={"script":"return document.querySelectorAll('iframe')[0];", "args":[]}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":{"element-6066-11e4-a52e-4f735466cecf":"57c69f02-ae5c-4234-992e-09525f059044"}}
_WD_ExecuteScript ==> Success [0]
```
you can test it with:
```autoit
Local $sResponse = '{"value":{"element-6066-11e4-a52e-4f735466cecf":"57c69f02-ae5c-4234-992e-09525f059044"}}'
Local $oJSON = Json_Decode($sResponse)
ConsoleWrite("! " & @CRLF)

Local $sValue = Json_Get($oJSON, '[value]')
ConsoleWrite("> "  & $sValue & @CRLF)
```
my result is:
```
+>Setting Hotkeys...--> Press Ctrl+Alt+F5 to Restart. --> Press Ctrl+Shift+F5 to Stop.
! 
> 
+>14:35:11 AutoIt3 ended. rc:0
```

Why I can't get value ?

**Comment by Danp2** [comment]
> Why I can't get value ?

Try this instead --

```autoit
#include "wd_core.au3"

Local $sResponse = '{"value":{"element-6066-11e4-a52e-4f735466cecf":"57c69f02-ae5c-4234-992e-09525f059044"}}'
Local $oJSON = Json_Decode($sResponse)
ConsoleWrite("! " & @CRLF)

Local $sValue = Json_Get($oJSON, $_WD_JSON_Element)
ConsoleWrite("> "  & $sValue & @CRLF)
```

**Comment by mlipok** [comment]
Thanks.
That do this trick.
image

**Comment by mlipok** [comment]
My tests show that there are differences
image

**Comment by Danp2** [comment]
> My tests show that there are differences

Going back to my earlier statement --

> In this case, a parent container refers to the parent element and any of its ancestral parent elements. This is why I said "But you can't just check the parent element if that's what you are thinking" because simply checking the parentElement's visibility isn't sufficient to ensure that the frame is visible.

Have you verified which one is providing the correct result?

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3:868]: > > > this won't be accurate when a parent container is hidden and not the actual frame.
> 
> > What you are refering here ?
> 
> In this case, `a parent container` refers to the parent element _and_ any of its ancestral parent elements. This is why I said "But you can't just check the parent element if that's what you are thinking" because simply checking the parentElement's visibility isn't sufficient to ensure that the frame is visible.

On the follwing screenshot frame `#19  null/13/0` is hidden thus all sub frames `null/13/0/.......`  in this case `#20...#27` are also hidden and only `WebDriver test` shows  IsHiddenWebDriver=True but `JS test` shows IsHiddenJS=False thus ... `IsHiddenJS` shows invalid values.

This is what you was thinking about ?

image

Comment (mlipok) [reply wd_helper.au3:868]: If so ... What I should do ? to leave only `IsHiddenWebDriver` feature and remove `IsHiddenJS` feature ?

**Comment by mlipok** [comment]
> > My tests show that there are differences
> 
> Going back to my earlier statement --
> 
> > In this case, a parent container refers to the parent element and any of its ancestral parent elements. This is why I said "But you can't just check the parent element if that's what you are thinking" because simply checking the parentElement's visibility isn't sufficient to ensure that the frame is visible.
> 
> Have you verified which one is providing the correct result?

take a look here: https://github.com/Danp2/au3WebDriver/pull/393#discussion_r1119840031

File: wd_helper.au3
Comment (Danp2) [reply wd_helper.au3:868]: I suggest removing `IsHiddenJS` if it isn't accurate.

Comment (mlipok) [reply wd_helper.au3]: Take a look again: [hidden status only from WebDriver - not JS](https://github.com/Danp2/au3WebDriver/pull/393/commits/bf60fce8e067e16a25b01ab6af054fe9ef433ce8)

**Comment by mlipok** [comment]
btw.
I think that we can close realted ISSUE: https://github.com/Danp2/au3WebDriver/issues/388

**Comment by Danp2** [comment]
@mlipok I added a link so that the issue will close once the PR is merged.

File: wd_helper.au3
Comment (mlipok) [reply wd_helper.au3]: Can we cosider this part as resolved ?

File: wd_demo.au3
Comment (Danp2) [inline wd_demo.au3:698]: This appears to be unfinished.

**Review by Danp2 — CHANGES_REQUESTED** [review]

File: wd_demo.au3
Comment (mlipok) [reply wd_demo.au3:698]: working on

**Comment by mlipok** [comment]
image

`#Region - Testing element location in frame set`
is related to planed `_WD_LocateElement()` function, which will be in separate PR soon.

File: wd_demo.au3
Comment (Danp2) [reply wd_demo.au3]: This still doesn't appear complete. Is Ex7 related to `_WD_LocateElement()`? If so, then why not submit it as part of that future PR instead of this one?

Comment (mlipok) [reply wd_demo.au3]: Checked. You have right.
Please take a look on: [removed Example 7](https://github.com/Danp2/au3WebDriver/pull/393/commits/9ed605ccc53fceb9d731b7aa482440a5041f104f)

**Review by Danp2 — APPROVED** [review]

**Comment by mlipok** [comment]
You can merge, which will let me do next PR with `_WD_LocateElement()`

**Pull request merged**
